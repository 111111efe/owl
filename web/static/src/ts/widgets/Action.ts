import { Widget } from "../core/widget";
import { Env } from "../env";
import { ActionStack } from "../services/action_manager";

export interface Props {
  stack: ActionStack;
}

export class Action extends Widget<Env, Props> {
  name = "action";
  template = `<div class="o_content"/>`;
  currentWidget: any;

  shouldUpdate(nextProps: Props) {
    if (nextProps.stack !== this.props.stack) {
      this.props = nextProps;
      this.setContentWidget();
    }
    return false;
  }

  async setContentWidget() {
    const info = this.props.stack[this.props.stack.length - 1];
    if (info && info.type === "client") {
      const Widget = info.Widget;
      let widget = new Widget(this, {});
      await widget.mount(this.el!);
      if (this.currentWidget) {
        this.currentWidget.destroy();
      }
      this.currentWidget = widget;
    }
  }
}
