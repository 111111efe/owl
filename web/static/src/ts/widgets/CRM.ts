import Widget from "../core/widget";
import { Env } from "../env";

const template = `
    <div class="o_crm">
        <span>CRM!!!!</span>
    </div>
`;

export default class Discuss extends Widget<Env> {
  name = "crm";
  template = template;
}
