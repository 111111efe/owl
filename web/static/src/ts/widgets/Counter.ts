import Widget from "../core/Widget";

const template = `
    <div>
        <button t-on-click="increment(-1)">-</button>
        <span style="font-weight:bold">Value: <t t-esc="state.counter"/></span>
        <button t-on-click="increment(1)">+</button>
    </div>
`;

export default class Counter extends Widget {
  name = "counter";
  template = template;
  state = {
    counter: 0
  };

  constructor(parent: Widget | null, props: {initialState?: number}) {
    super(parent);
    this.state.counter = props.initialState || 0;
  }

  increment(delta: number) {
    this.updateState({ counter: this.state.counter + delta });
  }
}
