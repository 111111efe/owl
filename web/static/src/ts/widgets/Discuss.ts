import { Widget } from "../core/widget";
import { Counter } from "./counter";
import { Env } from "../env";

const template = `
    <div class="o_discuss">
        <span>Root Widget</span>
        <button t-on-click="resetCounter">Reset</button>
        <button t-on-click="resetCounterAsync">Reset in 3s</button>
        <button t-on-click="toggle">Toggle  Counter</button>
        <input/>
        <t t-if="state.validcounter">
            <t t-widget="Counter" t-ref="counter" t-props="{initialState:4}"/>
        </t>
        <t t-else="1">
            <t t-widget="Counter" t-ref="counter" t-props="{initialState:7}"/>
        </t>
        <div ref="target"/>
    </div>
`;

export class Discuss extends Widget<Env> {
  name = "discuss";
  template = template;
  widgets = { Counter };
  state = { validcounter: true };

  resetCounter(ev: MouseEvent) {
    this.refs.counter.updateState({ counter: 3 });
  }

  resetCounterAsync(ev: MouseEvent) {
    setTimeout(() => {
      this.refs.counter.updateState({ counter: 3 });
    }, 3000);
  }

  toggle() {
    this.updateState({ validcounter: !this.state.validcounter });
  }
}
