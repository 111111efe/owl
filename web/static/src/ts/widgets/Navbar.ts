import { Widget } from "../core/widget";
import { Env, Menu } from "../env";

export interface Props {
  inHome: boolean;
}

export class Navbar extends Widget<Env, Props> {
  template = "navbar";

  getUrl(menu: Menu) {
    const action_id = String(menu.actionID);
    return this.env.router.formatURL("", { action_id });
  }

  toggleHome(ev: MouseEvent) {
    ev.preventDefault();
    this.trigger("toggle-home-menu");
  }
}
