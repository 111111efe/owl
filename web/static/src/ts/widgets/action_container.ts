import { ActionStack } from "../store/store";
import { Widget } from "./widget";

//------------------------------------------------------------------------------
// Types
//------------------------------------------------------------------------------

export interface Props {
  stack: ActionStack;
}

//------------------------------------------------------------------------------
// Action Container
//------------------------------------------------------------------------------

export class ActionContainer extends Widget<Props, {}> {
  template = "web.action_container";
  currentWidget: any;

  willStart() {
    return this.setContentWidget();
  }

  mounted() {
    if (this.currentWidget && this.currentWidget.el) {
      this.el!.appendChild(this.currentWidget.el);
      this.currentWidget.__mount();
    }
  }

  shouldUpdate(nextProps: Props) {
    if (nextProps.stack !== this.props.stack) {
      this.props = nextProps;
      this.setContentWidget();
    }
    return false;
  }

  async setContentWidget() {
    const info = this.props.stack[this.props.stack.length - 1];
    if (info && info.type === "client") {
      const Widget = info.Widget;
      let widget = new Widget(this, {});
      await widget.mount(this.el || document.createElement("div"));
      if (this.currentWidget) {
        this.currentWidget.destroy();
      }
      this.currentWidget = widget;
    }
  }
}
