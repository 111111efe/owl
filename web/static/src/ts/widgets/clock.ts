import { Widget } from "../core/widget";
import { Env } from "../env";

const template = `<div class="o_clock"><t t-esc="state.currentTime"/></div>`;

export class Clock extends Widget<Env, {}> {
  name = "clock";
  template = template;
  interval: any | undefined;

  state = {
    currentTime: ""
  };

  async willStart() {
    this.updateTime();
  }

  mounted() {
    this.interval = setInterval(this.updateTime.bind(this), 500);
  }

  willUnmount() {
    clearInterval(this.interval);
  }
  updateTime() {
    this.updateState({
      currentTime: new Date().toLocaleTimeString()
    });
  }
}
