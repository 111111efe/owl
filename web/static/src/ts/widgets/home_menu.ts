import { Widget } from "../core/widget";
import { Env } from "../env";

export class HomeMenu extends Widget<Env, {}> {
  template = "web.home_menu";
}
