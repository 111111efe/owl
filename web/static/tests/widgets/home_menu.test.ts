import { HomeMenu, Props } from "../../src/ts/widgets/home_menu";
import * as helpers from "../helpers";

//------------------------------------------------------------------------------
// Setup and helpers
//------------------------------------------------------------------------------

let fixture: HTMLElement;
let env: ReturnType<typeof helpers.makeTestEnv>;
let props: Props;
let templates: string;

beforeAll(async () => {
  templates = await helpers.loadTemplates();
});

beforeEach(() => {
  fixture = helpers.makeTestFixture();
  env = helpers.makeTestEnv();
  env.qweb.loadTemplates(templates);
  props = { menuInfo: helpers.makeDemoMenuInfo() };
});

afterEach(() => {
  fixture.remove();
});

//------------------------------------------------------------------------------
// Tests
//------------------------------------------------------------------------------

test("can be rendered", async () => {
  const homeMenu = new HomeMenu(env, props);
  await homeMenu.mount(fixture);
  expect(fixture.innerHTML).toMatchSnapshot();
});
