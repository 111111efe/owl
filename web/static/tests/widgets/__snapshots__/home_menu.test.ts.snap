// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can be rendered 1`] = `
"<div class=\\"o_home_menu\\">
    <div class=\\"o_apps\\">
      <a href=\\"#menu_id=96&amp;action_id=131\\" class=\\"o_app\\">
        <i class=\\"fa fa-comment o_app_icon fa-3x fa-fw\\"></i>
        <div class=\\"o_caption\\">
          Discuss
        </div>
      </a><a href=\\"#menu_id=205&amp;action_id=250\\" class=\\"o_app\\">
        <i class=\\"fa fa-pen o_app_icon fa-3x fa-fw\\"></i>
        <div class=\\"o_caption\\">
          Notes
        </div>
      </a><a href=\\"#menu_id=409&amp;action_id=597\\" class=\\"o_app\\">
        <i class=\\"fa fa-handshake o_app_icon fa-3x fa-fw\\"></i>
        <div class=\\"o_caption\\">
          CRM
        </div>
      </a>
    </div>
  </div>"
`;
